### PR TITLE
add support of feature flag for runnables

### LINKED ISSUE #4464: Run test hidden behind a feature by detecting the needed feature(s)
I have tests that require feature(s) to run
```rust
#[test]
#[cfg(feature = "serde1")]
fn test_some_test_of_serde_features() {
}
```
But when I run this, the executed task (not exactly this above):
```
> Executing task: cargo test --package rand --lib -- distributions::uniform::tests::test_serialization_uniform_duration --exact --nocapture <
```
where it should be 
```
> Executing task: cargo test --features serde1 --package rand --lib -- distributions::uniform::tests::test_serialization_uniform_duration --exact --nocapture <
```

I am unsure if this is possible, or if it has been raised before.

### PR DESCRIPTION
**Pull request by bnjjj**
close #4464

### CODE REVIEW

File: crates/ra_ide/src/runnables.rs [hunk 1]
```diff
@@ -160,11 +173,34 @@ fn runnable_mod(sema: &Semantics<RootDatabase>, module: ast::Module) -> Option<R
...
+        if let Some(features_needed) = cfg.minimal_features_needed() {
+            acc.extend(features_needed);
+        }
+        acc
+    });
+    if features_needed.is_empty() {
+        None
+    } else {
+        Some(features_needed)
+    }
```
Comment (Veetaha) [inline crates/ra_ide/src/runnables.rs:203]:
```suggestion
    Some(features_needed).filter(|it| !it.is_empty())
```

File: crates/ra_cfg/src/cfg_expr.rs [hunk 2]
```diff
@@ -33,6 +33,36 @@ impl CfgExpr {
...
+
+    /// Return minimal features needed
+    pub fn minimal_features_needed(&self) -> Option<Vec<SmolStr>> {
+        let mut features = vec![];
+        self.collect_minimal_features_needed(&mut features);
+        if features.is_empty() {
+            None
+        } else {
+            Some(features)
+        }
```
Comment (Veetaha) [inline crates/ra_cfg/src/cfg_expr.rs:45]: I don't why the vector must not be empty, but if this invariant does need to be hold it's better to encode this contract with a newtype:
```rust
struct Features(Vec<SmolStr>);
impl Features {
    fn new(feature_names: Vec<SmolStr>) -> Option<Self> {
        match feature_names.is_empty() {
            true => None,
            false => Self(feature_names)
        }
    }
}
```

Comment (bnjjj) [reply crates/ra_cfg/src/cfg_expr.rs]: I choose to return just a vector, you're right option here could be unnecessary.

File: crates/ra_cfg/src/cfg_expr.rs [hunk 3]
```diff
@@ -33,6 +33,33 @@ impl CfgExpr {
...
+    pub fn minimal_features_needed(&self) -> Vec<SmolStr> {
+        let mut features = vec![];
+        self.collect_minimal_features_needed(&mut features);
+
+        features
+    }
+
+    fn collect_minimal_features_needed(&self, features: &mut Vec<SmolStr>) {
+        match self {
+            CfgExpr::KeyValue { key, value } if key == "feature" => features.push(value.clone()),
```
Comment (matklad) [inline crates/ra_cfg/src/cfg_expr.rs:47]: `cfg` module by design doesn't know about cargo "features". So this code should live elsewhere, at the point where we have information about Cargo. I think `rust-analyzer` crate would be the most approporitate location.

File: crates/ra_ide/src/runnables.rs [hunk 4]
```diff
@@ -16,6 +16,7 @@ use std::fmt::Display;
 pub struct Runnable {
     pub range: TextRange,
     pub kind: RunnableKind,
+    pub features_needed: Option<Vec<SmolStr>>,
```
Comment (matklad) [inline crates/ra_ide/src/runnables.rs:19]: `features` is a cargo-specific concept. `ra_ide` layer doesn't know about `Cargo`, so it shouldn't talk about features directly. I think here we should store a `CfgExpr`, and then extract features out of it at the `rust-analyzer` layer.

**Comment by bnjjj** [comment]
Updated @matklad let me know if it's good for you :)

**Pull request merged**